File: 2016-2017/Java/Hw8/131044059/BigramMap.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
//package hw8;

import com.oracle.jrockit.jfr.DataType;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 *
 * @author segilmez
 * @param <T>
 */
public class BigramMap<T> implements Bigram<T> {

    private int dataType;
    private ArrayList datas = new ArrayList<>();

    private Map<Pair, Integer> bigrams = new HashMap();

    public BigramMap(int type) throws Exception {
        dataType = type;
        if (dataType < 1 || dataType > 3) {
            throw new Exception("Data type must be 1,2 or 3. ");
        }
    }

    private class Pair {

        public T first;
        public T second;

        public Pair(T t1, T t2) {
            first = t1;
            second = t2;
        }

        public Pair() {
            first = null;
            second = null;
        }

        @Override
        public int hashCode() {
            int ascii = 0;
            if (first instanceof String) {
                String str = ((String) first) + ((String) second);
                for (int i = 0; i < str.length(); i++) {
                    ascii += (int) str.charAt(i);
                }
            } else if (first instanceof Integer || first instanceof Double) {
                ascii = (Integer) first + (Integer) second;
            }

            return ascii;
        }

        @Override
        public boolean equals(Object obj) {

            if (obj == this) {
                return true;
            }
            Pair p = (Pair) obj;
            return (first.equals(p.first)) && second.equals(p.second);
        }

        @Override
        public String toString() {
            return first + " - " + second;
        }

    }

    /**
     *
     * @param fileName
     * @return
     * @throws Exception
     */
    @Override
    public boolean readFile(String fileName) throws Exception {
        try {
            Scanner scanner = new Scanner(new File(fileName));
            while (scanner.hasNext()) {
                datas.add(scanner.next());
            }
        } catch (Exception e) {
            throw new Exception("Unable open file");
        }

        T t1 = null, t2 = null;

        for (int i = 0; i < datas.size() - 1; i++) {
            if ((dataType == 1 && isInteger((String) datas.get(i)))
                    || (dataType == 3 && isDouble((String) datas.get(i)))
                    || dataType == 2) {
                Pair p = new Pair(t1, t2);
                p.first = (T) datas.get(i);
                p.second = (T) datas.get(i + 1);
                if (bigrams.containsKey(p)) {
                    bigrams.put(p, bigrams.get(p) + 1);
                } else {
                    bigrams.put(p, 1);
                }
            } else {
                throw new Exception("Data not acceptable.");
            }
        }

        return true;
    }

    @Override
    public int numGrams() {
        return datas.size() - 1;
    }

    @Override
    public int numOfGrams(T elem1, T elem2) {
        /*
        for(Pair p : bigrams.keySet()){
            System.out.println(p.first);
            if(p.first == elem1 || p.second == elem2)
                return bigrams.get(p);
        }
         */

        if (bigrams.containsKey(new Pair(elem1, elem2))) {
            return bigrams.get(new Pair(elem1, elem2));
        } else {
            return 0;
        }
    }

    @Override
    public String toString() {
        int num = datas.size();
        String str = "";
        while (num > 0) {
            for (Pair key : bigrams.keySet()) {
                if (bigrams.get(key) == num) {
                    str += key.toString() + " : " + num + "  |  ";
                }
            }

            num--;
        }
        return str;
    }

    public static boolean isInteger(String s) {
        try {
            Integer.parseInt(s);
        } catch (NumberFormatException | NullPointerException e) {
            return false;
        }
        return true;
    }

    public static boolean isDouble(String s) {
        try {
            Double.parseDouble(s);
        } catch (NumberFormatException | NullPointerException e) {
            return false;
        }
        return true;
    }

    public static void main(String[] args) throws IOException, Exception {
        BigramMap bg = new BigramMap(1);
        bg.readFile("t1.txt");
        System.out.println(bg.numGrams());
        System.out.println(bg.numOfGrams(1, 4));
        System.out.println(bg);
    }

}

File: 2016-2017/C++/Hw7/Hw7/BigramMap.cpp
#include "BigramMap.h"

template<class T>
BigramMap<T>::BigramMap(int type) {
    dataType = type;
    if(dataType <1 ||dataType>3)
       throw MyException("Data type must be 1 or 2. ");
}

template<class T>
BigramMap<T>::~BigramMap() {
}

template <class T>
bool BigramMap<T>::readFile(const string filename) {
    datas.clear();
    bigrams.clear();
    ifstream file; //file stream
    string str = ""; //each line 
    int i = 0;
    file.open(filename);
    if (!file.is_open()) {
        throw MyException("Unable to open file");
        return false;
    } else {
        while (file >> str) {
            if ((dataType == 1 && is_same<int, T>::value) ||
                    (dataType == 2 && is_same<string, T>::value) ||
                    (dataType == 3 && is_same<double, T>::value)) {
                elemCount++;
                continue;
            } else {
                throw "Data type not acceptable!\n";
                return false;
            }
        }
        file.close();
    }
    file.open(filename);
    if (!file.is_open()) {
        throw MyException("Unable to open file");
        return false;
    } else {
        T temp;
        while (file >> temp) {
            datas.push_back(temp);
        }
        if (elemCount == 0) {
            throw MyException("Empty file.");
            file.close();
            return false;
        } else if (datas.size() != elemCount || datas.size() < 2) {
            throw (MyException("Bad data."));
            file.close();
            return false;
        } 
            for (int i = 0; i < datas.size() - 1; i++) {
                pair<T, T> tempp;
                tempp.first = datas[i];
                tempp.second = datas[i + 1];
                setBg(tempp);
            }
            file.close();
        
    }
    return true;
}

template <class T>
int BigramMap<T>::numGrams()const {
    return this->elemCount - 1;
}

template <class T>
int BigramMap<T>::numOfGrams(T elem1, T elem2) {
    if (bigrams.find(pair<T, T> (elem1, elem2)) != bigrams.end()) {
        return this->bigrams[pair<T, T>(elem1, elem2)];
    } else
        return 0;
}

template <class T>
pair<T, T> BigramMap<T>::maxGrams() {
    int max = 0, i = 0;
    pair<T, T> p;
    for (int i = 0; i < datas.size() - 1; i++) {
        if (max < bigrams[pair<T, T>(datas[i], datas[i + 1])]) {
            max = bigrams[pair<T, T>(datas[i], datas[i + 1])];
            p.first = datas[i];
            p.second = datas[i + 1];
        }
    }
    return p;
}

template <class T>
void BigramMap<T>::print(std::ostream& output) {
    int num = elemCount;
    bool isCheck=false;
    while (num > 0) {
        for (int i = 0; i < datas.size() - 1; i++) {
            if (num == bigrams[std::pair<T, T>(datas[i], datas[i + 1])] && !isCheck) {
                output << datas[i] << " , " << datas[i + 1] << " : " << num << endl;
                isCheck=true;
            }
        }
        isCheck=false;
        num--;
    }
}


template <class T>
void BigramMap<T>::setBg(const pair<T, T> p) {
    if (bigrams.find(p) == bigrams.end()) {
        bigrams[p] = 1;
    } else {
        int count = bigrams[p];
        bigrams[p] = (count + 1);
    }
}

